test/cloudtest/test_k8s_node_recovery.py

```diff
@@ -0,0 +1,354 @@
+from dataclasses import dataclass
+from textwrap import dedent
+
+import pytest
+
+from materialize import ui
+from materialize.cloudtest.app.materialize_application import MaterializeApplication
+
+
+@dataclass
+class ReplicaDefinition:
+    cluster_name: str
+    index: int
+    availability_zone: str
+
+    def get_name(self) -> str:
+        return f"{self.cluster_name}_r_{self.index}"
+
+
+@dataclass
+class ClusterDefinition:
+    name: str
+    replica_definitions: list[ReplicaDefinition]
+
+    def create_replica_definitions_sql(self) -> str:
+        replica_definitions = []
+        for replica in self.replica_definitions:
+            replica_definitions.append(
+                f"{replica.get_name()} (SIZE = '1', AVAILABILITY ZONE '{replica.availability_zone}')"
+            )
+
+        return ", ".join(replica_definitions)
+
+
+def default_compute_cluster() -> ClusterDefinition:
+    compute_cluster = ClusterDefinition("c_compute", [])
+    compute_cluster.replica_definitions.append(
+        ReplicaDefinition(compute_cluster.name, 1, "1")
+    )
+    return compute_cluster
+
+
+def default_storage_cluster() -> ClusterDefinition:
+    storage_cluster = ClusterDefinition("c_storage", [])
+    storage_cluster.replica_definitions.append(
+        ReplicaDefinition(storage_cluster.name, 1, "1")
+    )
+    return storage_cluster
+
+
+REQUIRED_PROGRESS = 5
+
+
+def populate(
+    mz: MaterializeApplication,
+    compute_cluster: ClusterDefinition,
+    storage_cluster: ClusterDefinition,
+) -> None:
+    all_clusters = [compute_cluster, storage_cluster]
+
+    drop_cluster_statements = [
+        f"> DROP CLUSTER IF EXISTS {cluster.name};" for cluster in all_clusters
+    ]
+    # string needs the same indentation as testdrive script below
+    drop_cluster_statement_sql = "\n            ".join(drop_cluster_statements)
+
+    create_cluster_statements = [
+        f"> CREATE CLUSTER {cluster.name} REPLICAS ({cluster.create_replica_definitions_sql()});"
+        for cluster in all_clusters
+    ]
+    # string needs the same indentation as testdrive script below
+    create_cluster_statement_sql = "\n            ".join(create_cluster_statements)
+
+    mz.testdrive.run(
+        input=dedent(
+            f"""
+            > DROP MATERIALIZED VIEW IF EXISTS mv;
+            > DROP SOURCE IF EXISTS source;
+
+            {drop_cluster_statement_sql}
+            {create_cluster_statement_sql}
+
+            
+            > SELECT name FROM mz_clusters WHERE name IN ('{compute_cluster.name}', '{storage_cluster.name}');
+            {storage_cluster.name}
+            {compute_cluster.name}
+
+            > CREATE SOURCE source IN CLUSTER {storage_cluster.name}
+              FROM LOAD GENERATOR COUNTER
+              (TICK INTERVAL '500ms');
+
+            > SELECT COUNT(*) FROM (SHOW SOURCES) WHERE name = 'source';
+            1
+
+            > CREATE MATERIALIZED VIEW mv (f1) IN CLUSTER {compute_cluster.name} AS SELECT * FROM source;
```

Comment:
To avoid any optimizations now or in the future, consider creating the MV as `SELECT some_column + 1 FROM source`. This means a dataflow will surely be created, otherwise Mz may some day elect to not create a dataflow at all if the query is just `SELECT * FROM source`.